### PR TITLE
Linkcheck feature request

### PR DESCRIPTION
**Pull request by SG004**
I am submitting a pull request for Issue #8254 

Over the past week or so, I've looked into CircleCI and Sphinx. I had never touched them before, and I had a bit of trouble understanding how the Sphinx documents were actually generated. I see where the circleCI html-docs job creates the docker using the circleci/python:3.6 and I see how the build_docs command is run by calling setup.py, which then uses the setup.cfg file, but I get lost in how setup.py and setup.cfg interact.  I don't understand the error for why the html-docs job fails
snap of circleci html-docs job failure

### CODE REVIEW

**Comment by astrofrog** [comment]
Also it looks like CircleCI didn't actually run, hopefully that will be fixed when the commits are squashed

**Comment by SG004** [comment]
@bsipocz @astrofrog 
The issue I first talked about when I made the pull request happened when CircleCi ran. Is it failing due to some of the links being broken?

**Comment by pllim** [comment]
I don't see where the broken links are from the screenshot nor from the CircleCI log:
```
reading sources... [ 76%] api/astropy.visualization.AsymmetricPercentileInterval
reading sources... [ 76%] api/astropy.visualization.BaseInterval
reading sources... [ 76%] api/astropy.visualization.BaseStretch
Step was canceled
```

But maybe some suggestions at artefactual/archivematica-docs#165 would be relevant? You can opt to increase max retries, timeout limit, or flatly ignore a given URL.

**Comment by bsipocz** [comment]
@pllim - I'm not sure why the mess with the downloaded log, the linkchecker deifnitely run after that docs run. 

Anyway, here is one broken link: 

screen shot 2019-02-06 at 12 32 39

**Comment by pllim** [comment]
For that particular example, when I type it in manually, it redirects to https://journals.aas.org/manuscript-preparation/ . http://www.sphinx-doc.org/en/master/_modules/sphinx/builders/linkcheck.html seems to have logic for redirect. But somehow this case triggered 403 (forbidden) in CircleCI. :thinking:

**Comment by bsipocz** [comment]
@pllim - I suspect that restarting may solve it (redirects were picked up elsewhere). In either case it's actually good that it's broken as we need to think about how best to proceed anyway, I don't think it would do any good if such, most probably temporary, failures would break the docs build green status (yet I prefer to have the links checked).

So the dilemma, how to have this 🍰 and eat it, too.

**Comment by pllim** [comment]
Can linkcheck be a separate cron job? I think it is probably acceptable to only run linkcheck, say, weekly.

**Comment by bsipocz** [comment]
circleCI gurus need to comment on that, I'm not aware that we use circleCI cron anywhere.

Or do you suggest to add a docs building cron job to travis that also runs linkchecker? I rather like that solution.

**Comment by pllim** [comment]
I have to admit I don't remember the original motivation of moving doc build from Travis to CircleCI. Is creating a weekly doc build cron job on Travis does not defeat that original motivation, I have no objection.

**Comment by bsipocz** [comment]
It was mostly the artifact support for the individual commits. 

Frankly I think that anyone looking at the log of the cron jobs knows enough of what to look for in a failing sphinx build.

@astrofrog - What do you think about this plan?

**Comment by astrofrog** [comment]
I use the link checker for glue and I just have it as a normal check and it's quite rare that it will fail for no reason. I would personally suggest for now adding it to the CircleCI build (which we have there because as @bsipocz said we can then upload the built docs to the artifacts and have Giles provide a link to preview the built docs in the comments here). I personally always forget to check cron-only jobs in the CI so I'm not sure that would really help. I find it more helpful if it fails and I then can go and make a decision as to whether to ignore the failure or not before merging a PR rather than have to always go to the CI manually to check what the cron job is doing.

Note that there are two config items that can make things more robust - ``linkcheck_timeout`` can be increased to make sure we don't time out too soon, ``linkcheck_ignore`` can be used to ignore specific domains that are not reliable, and ``linkcheck_retry`` can be used to try more than once (defaults to 1). So I'd recommend trying those first and if we really do have too many failures we can reconsider? See all the options at http://www.sphinx-doc.org/en/1.6/config.html#options-for-the-linkcheck-builder

**Comment by SG004** [comment]
@astrofrog I will try to do various combinations of the `linkcheck_timeout` and `linkcheck-retry` to see if we can get the circleCI build to complete with any problems.

**Comment by astrofrog** [comment]
Some of the failures are real by the way, don't assume the links should all be working right now :)

**Comment by SG004** [comment]
I will continue to try and test different combinations to see if there is a decrease in the number of broken links. If there is no difference, I will revert the commit back to the default linkcheck configuration. Let me know what I should do from that point

**Comment by pllim** [comment]
The links that persistently break need to be investigated. Listing them out in a comment will be a good start. Thank you!

**Comment by SG004** [comment]
The first two attempts yielded 23 broken links. Those attempts being 'default', and '2 retries, 10 second timeout'
The third attempt the '3 retries and 20 second timeout' yielded only 20 broken links.

[Link check 3 retries 20 second timeout.docx](https://github.com/astropy/astropy/files/2847239/Link.check.3.retries.20.second.timeout.docx)
broken links

**Comment by bsipocz** [comment]
Could you copy paste the broken ones in a gist or even here in a comment? docx files are not really suitable for this kind of stuff, neither should anyone open such files downloaded from anywhere on the internet.

**Comment by SG004** [comment]
My Apologies, here is a gist of the broken links
https://gist.github.com/SG004/567493b1f0cc69558abf407ddb0bfeab

**Comment by pllim** [comment]
Thanks for compiling the list! I'll have a look and provide suggestions where I can.

p.s. I think there is *still* a bug over at Gist where comments there do not trigger notifications to people involved, so I would like to keep all the conversation here instead of over at Gist page.

**Comment by pllim** [comment]
#### Suggestions on how to fix the broken links

1. http://avesbiodiv.mncn.csic.es/estadistica/ejemploaic.pdf -- Either @mirca can provide an alternate link (he added the original reference in #4716) or we remove it altogether.

2. http://journals.aas.org/authors/manuscript.html -- Replace with https://journals.aas.org/manuscript-preparation/

3. http://nbviewer.ipython.org/urls/ -- This is actually part of a URL accidentally broken into two in the docstring.  http://nbviewer.ipython.org/urls/gist.github.com/ChrisBeaumont/5758381/raw/descriptor_writeup.ipynb should all be in one line.

4. https://github.com/numpy/numpy/blob/master/doc/HOWTO_BUILD_DOCS.rst.txt -- I asked over at numpy/numpy#12954

5. https://setuptools.readthedocs.iosetuptools.html#automatic-script-creation -- Replace with https://setuptools.readthedocs.io/en/latest/setuptools.html#automatic-script-creation

6. https://github.com/astropy/pytest-mpl/blob/master/README.rst -- Replace with https://github.com/matplotlib/pytest-mpl/blob/master/README.rst (tsk tsk @astrofrog )

7. https://www.anaconda.com/blog/developer-blog/conda-data-science/ -- Replace with https://www.anaconda.com/category/developer-blog/ (looks like Anaconda redesigned their blog layout)

8. http://www.stsci.edu/institute/software_hardware/pyfits -- Replace with https://github.com/spacetelescope/pyfits

9. (a) is Same as **8**; (b) http://listmgr.cv.nrao.edu/pipermail/fitsbits/2002-April/001094.html -- Remove the mention of it.

(I don't want to lose my text so far, so I am posting this incomplete list. I will add more as follow-up edits.)

**Comment by grantjenks** [comment]
Re: (13) above: Sorry, no it is not.

**Comment by astrofrog** [comment]
(16) -> https://astropy.stsci.edu/data/l1448/l1448_13co.fits

(17) -> the docs for v0.2.5 were never actually built. Why/where are we linking to this?

**Comment by pllim** [comment]
@astrofrog -- (17) In https://raw.githubusercontent.com/astropy/astropy/master/docs/whatsnew/0.2.rst

**Comment by astrofrog** [comment]
Ok - I'd just recommend using http://docs.astropy.org/en/older-docs-archive/v0.2.4/whatsnew/0.2.html instead

**Comment by bsipocz** [comment]
> @astrofrog -- (17) In https://raw.githubusercontent.com/astropy/astropy/master/docs/whatsnew/0.2.rst

double backtick all of these?

**Comment by mirca** [comment]
> 1. http://avesbiodiv.mncn.csic.es/estadistica/ejemploaic.pdf -- Either @mirca can provide an alternate link (he added the original reference in #4716) or we remove > it altogether. @bsipocz voted to remove it altogether.

I agree it can be removed altogether.

**Comment by SG004** [comment]
Should I create a new pull request in a new branch with all of the replacements and removals for the links, or should I just make the changes in this PR?

**Comment by bsipocz** [comment]
@SG004 - No, those changes should be part of this PR.

File: astropy/stats/info_theory.py [hunk 1]
```diff
@@ -391,8 +391,6 @@ def akaike_info_criterion_lsq(ssr, n_params, n_samples):
 
     References
     ----------
-    .. [1] Akaike Information Criteria
```
Comment (pllim) [inline astropy/stats/info_theory.py]: I think if you remove `[1]` here, then `[2]` now becomes `[1]` and so on?

**Comment by pllim** [comment]
@SG004 , do you have plans to follow up on the rest of the suggestions? If no, do you mind if I push to this PR?

**Comment by SG004** [comment]
I do plan to follow up. There were a few of the suggestions from your list comment @pllim , I wasn't sure we're fully resolved

**Comment by pllim** [comment]
@SG004 , except for (10) and (15), everything else has a solution. Please let me know if anything is still unclear. Thanks!

**Comment by pllim** [comment]
UPDATE: @SG004 , I have updated the suggestions above with solutions for all the items. Thanks for your patience!

File: docs/install.rst [hunk 2]
```diff
@@ -407,6 +407,7 @@ install sphinx-astropy with conda, graphviz will automatically get installed,
 but if you use pip, you will need to install Graphviz separately as it isn't
 a Python package.
 
+.. _astropy-doc-building:
```
Comment (pllim) [inline docs/install.rst]: This might need a blank line after. I am not 100% sure.

File: docs/table/masking.rst [hunk 3]
```diff
@@ -28,9 +28,9 @@ on masked arrays
    convention of ignoring masked (invalid) values.  This differs from
    the behavior of the floating point ``NaN``, for which the sum of an
    array including one or more ``NaN's`` will result in ``NaN``.
-   See `<http://www.numpy.org/NA-overview.html>`_ for a very
-   interesting discussion of different strategies for handling
-   missing data in the context of `numpy`.
+
+   See `<https://www.numpy.org/neps/>`_ for information on NumPy
```
Comment (pllim) [inline docs/table/masking.rst]: Please add mention to the specific NEP numbers, as indicated.

File: docs/whatsnew/0.2.rst [hunk 4]
```diff
@@ -4,6 +4,5 @@
 What's New in Astropy 0.2
 *************************
 
-See this page in the `Astropy v0.2 documentation`__.
+``See this page in the `Astropy v0.2 documentation`__.``
```
Comment (pllim) [inline docs/whatsnew/0.2.rst]:
```suggestion
See this page in the ``Astropy v0.2 documentation``.
```

File: docs/whatsnew/0.3.rst [hunk 5]
```diff
@@ -4,6 +4,5 @@
 What's New in Astropy 0.3?
 **************************
 
-See this page in the `Astropy v0.3 documentation`__.
+``See this page in the `Astropy v0.3 documentation`__.``
```
Comment (pllim) [inline docs/whatsnew/0.3.rst]:
```suggestion
See this page in the ``Astropy v0.3 documentation``.
```

Comment (SG004) [reply docs/whatsnew/0.3.rst]: I have addressed this in the most recent commit.

File: docs/whatsnew/0.2.rst (hunk 4 above)
Comment (SG004) [reply docs/whatsnew/0.2.rst]: I have addressed this in the most recent commit.

File: docs/table/masking.rst (hunk 3 above)
Comment (SG004) [reply docs/table/masking.rst]: I have addressed this in the most recent commit.

File: docs/install.rst (hunk 2 above)
Comment (SG004) [reply docs/install.rst]: I have addressed this in the most recent commit.

**Comment by pllim** [comment]
Well, I don't understand why this is still broken on link check...
```
(line    5) broken    https://journals.aas.org/manuscript-preparation/ - 403 Client Error: Forbidden for url: https://journals.aas.org/manuscript-preparation/
writing output... [ 62%] api/astropy.units.format.OGIP
```

**Comment by SG004** [comment]
There are 3 remaining broken links. The second one listed below is new and I will list them below:

**1. writing output... [ 62%] api/astropy.units.format.LatexInline**
   (line    5) broken    https://journals.aas.org/manuscript-preparation/ - 403 Client Error: Forbidden for 
   url: https://journals.aas.org/manuscript-preparation/

**2. writing output... [ 85%] development/docrules**
   (line   76) broken    http://docutils.sourceforge.net/rst.html - 
   HTTPConnectionPool(host='docutils.sourceforge.net', port=80): Max retries exceeded with url: /rst.html 
   (Caused by ConnectTimeoutError(<urllib3.connection.HTTPConnection object at 0x7f1665acfcc0>, 
   'Connection to docutils.sourceforge.net timed out. (connect timeout=20)'))

**3. writing output... [ 96%] units/format**
    (line  155) broken    https://journals.aas.org/manuscript-preparation/ - 403 Client Error: Forbidden for 
    url: https://journals.aas.org/manuscript-preparation/

**Comment by pllim** [comment]
Re: https://github.com/astropy/astropy/pull/8386#issuecomment-464241545

(1) and (3) -- Something is weird between Travis CI and aas.org blocking it. We might just have to exclude this one from the check.

(2) looks like a temporarily thing. The link works now and the CI didn't catch it before. So it'll probably go away in the next run.

One more thing that @bsipocz brought up but is not addressed yet is moving the linkcheck to cron job. That should be the last step because we want to make sure at least (1) and (3) are handled in the PR's CI before doing that.

**Comment by SG004** [comment]
> There are 3 remaining broken links. The second one listed below is new and I will list them below:
> 
> **1. writing output... [ 62%] api/astropy.units.format.LatexInline**
> (line 5) broken https://journals.aas.org/manuscript-preparation/ - 403 Client Error: Forbidden for
> url: https://journals.aas.org/manuscript-preparation/
> 
> **2. writing output... [ 85%] development/docrules**
> (line 76) broken http://docutils.sourceforge.net/rst.html -
> HTTPConnectionPool(host='docutils.sourceforge.net', port=80): Max retries exceeded with url: /rst.html
> (Caused by ConnectTimeoutError(<urllib3.connection.HTTPConnection object at 0x7f1665acfcc0>,
> 'Connection to docutils.sourceforge.net timed out. (connect timeout=20)'))
> 
> **3. writing output... [ 96%] units/format**
> (line 155) broken https://journals.aas.org/manuscript-preparation/ - 403 Client Error: Forbidden for
> url: https://journals.aas.org/manuscript-preparation/
> 
> How should we address these issues? I think we can fix the 2nd broken link by increasing the connect timeout. @pllim @astrofrog

After changing the linkcheck settings to try and resolve the second broken link, a new link became broken

writing output... [ 70%] api/astropy.utils.decorators.lazyproperty
(line   33) broken    http://code.activestate.com/recipes/363602-lazy-property-evaluation - HTTPConnectionPool(host='code.activestate.com', port=80): Max retries exceeded with url: /recipes/363602-lazy-property-evaluation (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f29f2a19860>: Failed to establish a new connection: [Errno 110] Connection timed out',))

**Comment by SG004** [comment]
> Re: [#8386 (comment)](https://github.com/astropy/astropy/pull/8386#issuecomment-464241545)
> 
> (1) and (3) -- Something is weird between Travis CI and aas.org blocking it. We might just have to exclude this one from the check.
> 
> (2) looks like a temporarily thing. The link works now and the CI didn't catch it before. So it'll probably go away in the next run.
> 
> One more thing that @bsipocz brought up but is not addressed yet is moving the linkcheck to cron job. That should be the last step because we want to make sure at least (1) and (3) are handled in the PR's CI before doing that.

html-docs now passes after ignoring the aas link, but for some reason there is now a merge conflict on the conf.py file. 
@pllim

**Comment by pllim** [comment]
You would have to rebase. Please don't click the "resolve conflicts" button.

**Comment by SG004** [comment]
> You would have to rebase. Please don't click the "resolve conflicts" button.

rebase to what?

**Comment by pllim** [comment]
```
git fetch upstream master
git rebase upstream/master
```

If there is somehow a conflict, you need to manually resolve it then

```
git add filename
git rebase --continue
```

When it is done, you will have to do a force push back out.

**Comment by pllim** [comment]
If you are worried that the rebase would go wrong, please make a backup branch first.

**Comment by pllim** [comment]
Ops looks like you also needed this step
```
git submodule update
```
Sorry for the oversight!

**Comment by SG004** [comment]
> Ops looks like you also needed this step
> 
> ```
> git submodule update
> ```
> Sorry for the oversight!

When I enter in the ```git submodule update``` command, nothing happens.

**Comment by pllim** [comment]
`git submodule update` should silently update the `astropy-helpers` so it is in-sync with what is on `master`. 🤔

**Comment by SG004** [comment]
> `git submodule update` should silently update the `astropy-helpers` so it is in-sync with what is on `master`. 🤔

It isn't working, and I am not quite sure what to do. I did create the backup branch, should I open a different PR with a reference to this one, using the backup branch?

**Comment by bsipocz** [comment]
@SG004 - no need, let's resolve this instead. It's a fairly typical issue, so learning how to deal with it pays of long term.
Are you on the astropy slack?

**Comment by SG004** [comment]
> @SG004 - no need, let's resolve this instead. It's a fairly typical issue, so learning how to deal with it pays of long term.
> Are you on the astropy slack?

I am not currently on the astropy slack. How would I go about joining?

**Comment by bsipocz** [comment]
http://joinslack.astropy.org/

**Comment by bsipocz** [comment]
This is all green now, so I'm merging. 

Very good job @SG004, thank you!

File: .circleci/config.yml [hunk 6]
```diff
@@ -106,6 +106,9 @@ jobs:
             LC_CTYPE: C
             LC_ALL: C
             LANG: C
+      - run:
```
Comment (pllim) [inline .circleci/config.yml]: @bsipocz , are we moving this to a cron job in a separate PR?

**Comment by pllim** [comment]
Thanks for your patience, @SG004 !

File: .circleci/config.yml (hunk 6 above)
Comment (bsipocz) [reply .circleci/config.yml]: There is no circleCI cron, afaik, so no we need to rely on to have it run for the PRs. Tom was very optimistic that it won't fail (https://github.com/astropy/astropy/pull/8386#issuecomment-461344945) so we may give it a go this way (and if it's flaky then move the whole stuff to travis cron)

Comment (pllim) [reply .circleci/config.yml]: Ah, I must have missed that conclusion. Thanks for the clarification!

**Pull request merged**